Feature/general refactor

jenniferlynparsons opened the pull request.

jenniferlynparsons committed:
rest api refactor

jenniferlynparsons committed:
clean up action errors and move id creation to server side

jenniferlynparsons committed:
move brewtime selector out of teadetails

jenniferlynparsons committed:
add selector for tea types

jenniferlynparsons committed:
refactor tea type and tea server error messages and validation

jenniferlynparsons committed:
refactor validation and clean up tests

jenniferlynparsons committed:
cleanup flash notifications and input error messages

jenniferlynparsons committed:
update flash functionality

jenniferlynparsons committed:
updated flash and action promises on editors

jenniferlynparsons committed:
Revert "loading action, reducer, tests and teatype loader"

This reverts commit adacbf7ad55f6fa5800887f746e9ff881b8ab669.

jenniferlynparsons committed:
loading action, reducer, tests and teatype loader

JeffreyATW commented on client/src/actions/authActions.js:
```diff
@@ -1,25 +1,28 @@
-        const errorObj = error.response.data;
-        dispatch(authActions.setErrorResponse(errorObj));
-      });
+      .catch(error => console.log(error));
```
Can be simplified to `.catch(console.log)` (for all occurrences).

JeffreyATW commented on client/src/components/FormComponents/InputField/InputField.jsx:
```diff
@@ -8,7 +8,9 @@ class InputField extends React.Component {
             data-testid={this.props.datatestid}
             className={
-              !this.props.error ? this.props.errorClass : this.props.className
+              !this.props.inputValidation
```
I feel like `valid` would be a better name than `inputValidation`.

JeffreyATW commented on server/routes/api/users.js:
```diff
@@ -27,9 +27,7 @@ router.post("/register", (req, res) => {
-      return res
-        .status(400)
-        .json({ emailAlreadyExists: "Email already exists" });
+      return res.json({ emailDoesNotExist: "Email already exists" });
```
I understand that you switched your validation where `false` was once `true`, but this line doesn't make any sense. I would rename the key to `duplicateEmail`. It will make less sense in the cases of `inputValidation.duplicateEmail = false`, but it is easier to understand when you see how all other `inputValidation` values are `true` when valid and `false` when invalid.

JeffreyATW commented on server/routes/api/teas.js:
```diff
@@ -21,11 +22,11 @@ router.post("/new-tea", (req, res) => {
   Tea.findOne({ id: req.body.id }).then(tea => {
     if (tea) {
-      return res.status(400).json({ teaConflict: "This tea already exists" });
+      return res.json({ noDuplicate: "This tea already exists" });
```
Similarly, would suggest renaming this to `duplicateTea` or just `duplicate`.

jenniferlynparsons committed:
local state loading status

jenniferlynparsons committed:
integrate suggested changes from code review

jenniferlynparsons committed:
Merge branch 'feature/general-refactor' into feature/loading-state-setup

jenniferlynparsons committed:
refinement and update tests

jenniferlynparsons committed:
Merge branch 'feature/general-refactor' into feature/loading-state-setup

jenniferlynparsons committed:
resolved conflict issues

jenniferlynparsons committed:
loading status fix

jenniferlynparsons committed:
realign inputValidation and valid names

jenniferlynparsons committed:
server error cleanup

jenniferlynparsons committed:
Merge branch 'feature/server-error-tests' into feature/general-refactor

jenniferlynparsons committed:
remove server error tests

separate branch created to preserve the work done. ticket create to finish the work.

jenniferlynparsons committed:
change componentwillrecieve to componentdidupdate

jenniferlynparsons committed:
improved flash messages

JeffreyATW commented on server/routes/api/teas.js:
```diff
@@ -40,7 +41,10 @@ router.post("/new-tea", (req, res) => {
+// @route PUT api/teas
+// @desc Update tea
+// @access Public
+router.put("/", (req, res) => {
```
Did you decide to keep the ID out of this `PUT` path, or was this an oversight? REST spec states you should be getting ID from `params`, not `body`.

JeffreyATW commented on server/routes/api/teas.js:
```diff
@@ -87,7 +96,10 @@ router.get("/tea/:id", (req, res) => {
+// @route GET api/teas
+// @desc Get all teas
+// @access Public
+router.get("/:id", function(req, res) {
```
The `id` here should refer to the Tea ID, not a user's ID.

JeffreyATW commented on server/routes/api/teas.js:
```diff
@@ -73,7 +80,9 @@ router.delete("/delete-tea/:id", (req, res) => {
+// @route GET api/teas/:id
+// @desc Get individual tea
+// @access Public
 router.get("/tea/:id", (req, res) => {
```
The full path here is `/api/teas/tea/:id` - the `teas/tea` part seems redundant to me.

JeffreyATW commented on server/routes/api/tea-types.js:
```diff
@@ -77,8 +82,10 @@ router.delete("/delete-tea-type/:id", (req, res) => {
+// @route GET api/tea-types/:id
+// @desc Get individual tea type
+// @access Public
+router.get("/type/:id", (req, res) => {
```
Same issue here - `tea-types/types` is redundant.

JeffreyATW commented on server/routes/api/tea-types.js:
```diff
@@ -91,8 +98,14 @@ router.get("/tea-type/:id", (req, res) => {
+// @route GET api/tea-types/:id
+// @desc Get all tea types
+// @access Public
+router.get("/:id", function(req, res) {
```
`id` should refer to the tea type, not a user.

JeffreyATW commented on client/src/components/Auth/Login/Login.jsx:
```diff
@@ -11,19 +11,16 @@ export class Login extends React.Component {
-            className="notification is-danger"
-            data-testid="incompletenotice"
-          >
+        {!this.props.inputValidation.complete && (
```
I don't think you should store a separate flag for whether the rest of the fields are valid or not. This is data duplication and is at risk of becoming out of sync. I would call a function that returns a boolean based on the states of all other fields in the form instead.

JeffreyATW commented on client/src/components/FormComponents/InputField/InputField.jsx:
```diff
@@ -8,7 +8,9 @@ class InputField extends React.Component {
-              !this.props.error ? this.props.errorClass : this.props.className
+              !this.props.valid
+                ? this.props.className + " " + this.props.errorClass
+                : this.props.className
```
Nitpick: I would simplify this to:
```
className={[this.props.className, this.props.valid ? '' : this.props.errorClass].join(' ')]}
```

JeffreyATW commented on client/src/components/Tea/TeaEditor/TeaEditor.jsx:
```diff
@@ -78,31 +78,32 @@ export class TeaEditor extends React.Component {
-                  !this.props.errors.teaType ? "select is-danger" : "select"
+                  !this.props.inputValidation.teaType
+                    ? "select is-danger"
+                    : "select"
```
Similar nitpick: This could be
```
className={['select', this.props.inputValidation.teaType ? '' : 'is-danger'].join(' ')}
```

jenniferlynparsons committed:
convert teatype to id instead of name

jenniferlynparsons committed:
refactor tea flash

jenniferlynparsons committed:
refine REST url paths

jenniferlynparsons committed:
refine rest url again

jenniferlynparsons committed:
Merge branch 'feature/loading-state-setup' into feature/general-refactor

jenniferlynparsons committed:
merged in loading state

redux render test wip

jenniferlynparsons committed:
moved redux rendered tests to a separate ticket

jenniferlynparsons committed:
Revert "moved redux rendered tests to a separate ticket"

This reverts commit 734fc9b12e7f730f71bb7173748b66300e0eb264.

jenniferlynparsons committed:
refine loading message display

jenniferlynparsons committed:
updated tests

jenniferlynparsons committed:
updated tests

jenniferlynparsons committed:
removed connected render tests

jenniferlynparsons committed:
added loading state to the rest of the components

jenniferlynparsons committed:
minor cleanup

jenniferlynparsons committed:
clean up input validation

jenniferlynparsons committed:
fix tea api route

jenniferlynparsons committed:
moved input validation to helper func

jenniferlynparsons committed:
validation cleanup

jenniferlynparsons committed:
fix initial empty time on timer

jenniferlynparsons committed:
refine timer mount

jenniferlynparsons committed:
refine validation helper

JeffreyATW commented on client/src/lib/validationComplete.js:
```diff
@@ -0,0 +1,11 @@
+  });
+
+  return valid;
+};
```
Does this apply for all forms? You probably want to have a custom validation function for every form.

On that note, there doesn't seem to be much of a reason why this needs to be passed as a prop. If you're going to use a single validation functions for all fields, why not just import it directly into the component file?

jenniferlynparsons replied:
:+1: for moving the function into the component. 

what would the advantage be to a separate function for each form?

JeffreyATW replied:
I forgot that this checks `inputValidation`, not the actual field values themselves. I guess if something lives in the `inputValidation` object, then if any of them are false, this function should return false, so this would be fine in that case...

You might find certain situations where one field is invalid but it's okay because another field is not (e.g. where you need to fill in one OR the other). You could potentially create a form-specific validation function in that case, though.

JeffreyATW replied:
BTW, you can rewrite this file like:
```
export const validationComplete = inputs => Object.values(inputs).every(Boolean);
```
This starts by creating an array of the values (since the key names don't matter) and determining whether all of them are truthy or not. The `Boolean` constructor is essentially an identity function that does this: `x => !!x` which is perfect for passing into the `every` method.

jenniferlynparsons committed:
moved validation function to component

jenniferlynparsons committed:
updates validation function
